Disallow BuildServices to be used as inputs to ValueSources obtained at configuration time

A `ValueSource` can take parameters in the form of Properties. A `BuildService` is obtained as a `Provider`. This `Provider` can be wired to the input `Property` of the ValueSource. It works well without the configuration cache. However, when the configuration cache is enabled and the ValueSource is obtained at the configuration time, things go wrong. The cache cannot be reused, as the fingerprinting routine cannot restore the `BuildService`; the build fails with a cryptic exception.

### Expected Behavior
Either of two options are expected:
1. If Gradle is not going to support this use case: provide an error with a clear message that this use is unsupported, preferably when storing the cache
2. Otherwise, this should just work, and the service instance should be available for the ValueSource computation.

### Current Behavior
The build fails when trying to check the cache validity:
```
org.gradle.api.GradleException: Could not load the value of field `__service__` of `ServiceValueSource$Params` bean found in field `parameters` of `org.gradle.api.internal.provider.DefaultValueSourceProviderFactory$DefaultObtainedValue` bean found in field `obtainedValue` of `org.gradle.configurationcache.fingerprint.ConfigurationCacheFingerprint$ValueSource` bean found in Gradle runtime.
  	at org.gradle.configurationcache.serialization.beans.BeanPropertyReaderKt.readPropertyValue(BeanPropertyReader.kt:108)
  	at org.gradle.configurationcache.serialization.beans.BeanPropertyReader.readStateOf(BeanPropertyReader.kt:67)
  	at org.gradle.configurationcache.serialization.codecs.BeanCodec.readBeanOf(BeanCodec.kt:72)
  	at org.gradle.configurationcache.serialization.codecs.BeanCodec.decode(BeanCodec.kt:47)
  	at org.gradle.configurationcache.serialization.CombinatorsKt$reentrant$1$decodeLoop$1.invokeSuspend(Combinators.kt:165)
  	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
  	at kotlin.coroutines.ContinuationKt.startCoroutine(Continuation.kt:115)
  	at org.gradle.configurationcache.serialization.CombinatorsKt$reentrant$1.decodeLoop(Combinators.kt:166)
  	at org.gradle.configurationcache.serialization.CombinatorsKt$reentrant$1.decode(Combinators.kt:130)
  	at org.gradle.configurationcache.serialization.codecs.BindingsBackedCodec.decode(BindingsBackedCodec.kt:59)
  	at org.gradle.configurationcache.serialization.DefaultReadContext.read(Contexts.kt:259)
  	at org.gradle.configurationcache.fingerprint.ConfigurationCacheFingerprintChecker.checkBuildScopedFingerprint(ConfigurationCacheFingerprintChecker.kt:58)
  	at org.gradle.configurationcache.fingerprint.ConfigurationCacheFingerprintController.checkBuildScopedFingerprint(ConfigurationCacheFingerprintController.kt:243)
  	at org.gradle.configurationcache.DefaultConfigurationCache$checkBuildScopedFingerprint$1.invokeSuspend(DefaultConfigurationCache.kt:461)
  	at org.gradle.configurationcache.DefaultConfigurationCache$readFingerprintFile$1$1.invokeSuspend(DefaultConfigurationCache.kt:480)
  	at org.gradle.configurationcache.ConfigurationCacheIO$withReadContextFor$1$1$1$1.invokeSuspend(ConfigurationCacheIO.kt:248)
  	at org.gradle.configurationcache.serialization.RunningKt$runReadOperation$2.invokeSuspend(Running.kt:34)
  	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
  	at kotlin.coroutines.ContinuationKt.startCoroutine(Continuation.kt:115)
  	at org.gradle.configurationcache.serialization.RunningKt.runToCompletion(Running.kt:56)
  	at org.gradle.configurationcache.serialization.RunningKt.runReadOperation(Running.kt:33)
  	at org.gradle.configurationcache.ConfigurationCacheIO.withReadContextFor$configuration_cache(ConfigurationCacheIO.kt:247)
  	at org.gradle.configurationcache.DefaultConfigurationCache.readFingerprintFile(DefaultConfigurationCache.kt:478)
  	at org.gradle.configurationcache.DefaultConfigurationCache.checkBuildScopedFingerprint(DefaultConfigurationCache.kt:459)
  	at org.gradle.configurationcache.DefaultConfigurationCache.checkFingerprintAgainstLoadedProperties(DefaultConfigurationCache.kt:441)
  	at org.gradle.configurationcache.DefaultConfigurationCache.checkFingerprint(DefaultConfigurationCache.kt:429)
  	at org.gradle.configurationcache.DefaultConfigurationCache.access$checkFingerprint(DefaultConfigurationCache.kt:53)
  	at org.gradle.configurationcache.DefaultConfigurationCache$checkFingerprint$1.invoke(DefaultConfigurationCache.kt:278)
  	at org.gradle.configurationcache.DefaultConfigurationCache$checkFingerprint$1.invoke(DefaultConfigurationCache.kt:271)
  	at org.gradle.configurationcache.ConfigurationCacheRepository$StoreImpl$useForStateLoad$2.invoke(ConfigurationCacheRepository.kt:164)
  	at org.gradle.configurationcache.ConfigurationCacheRepository$StoreImpl$useForStateLoad$2.invoke(ConfigurationCacheRepository.kt:163)
  	at org.gradle.configurationcache.ConfigurationCacheRepository$withExclusiveAccessToCache$1.create(ConfigurationCacheRepository.kt:244)
  	at org.gradle.cache.internal.LockOnDemandCrossProcessCacheAccess.withFileLock(LockOnDemandCrossProcessCacheAccess.java:90)
  	at org.gradle.cache.internal.DefaultCacheAccess.withFileLock(DefaultCacheAccess.java:191)
  	at org.gradle.cache.internal.DefaultPersistentDirectoryStore.withFileLock(DefaultPersistentDirectoryStore.java:188)
  	at org.gradle.cache.internal.DefaultCacheFactory$ReferenceTrackingCache.withFileLock(DefaultCacheFactory.java:209)
  	at org.gradle.configurationcache.ConfigurationCacheRepository.withExclusiveAccessToCache(ConfigurationCacheRepository.kt:242)
  	at org.gradle.configurationcache.ConfigurationCacheRepository.access$withExclusiveAccessToCache(ConfigurationCacheRepository.kt:46)
  	at org.gradle.configurationcache.ConfigurationCacheRepository$StoreImpl.useForStateLoad(ConfigurationCacheRepository.kt:163)
  	at org.gradle.configurationcache.DefaultConfigurationCache.checkFingerprint(DefaultConfigurationCache.kt:271)
  	at org.gradle.configurationcache.DefaultConfigurationCache.determineCacheAction(DefaultConfigurationCache.kt:221)
  	at org.gradle.configurationcache.DefaultConfigurationCache.initializeCacheEntry(DefaultConfigurationCache.kt:114)
  	at org.gradle.configurationcache.ConfigurationCacheBuildTreeLifecycleControllerFactory.createRootBuildController(ConfigurationCacheBuildTreeLifecycleControllerFactory.kt:54)
  	at org.gradle.composite.internal.DefaultRootBuildState.<init>(DefaultRootBuildState.java:73)
  	at org.gradle.composite.internal.BuildStateFactory.createRootBuild(BuildStateFactory.java:66)
  	at org.gradle.composite.internal.DefaultIncludedBuildRegistry.createRootBuild(DefaultIncludedBuildRegistry.java:91)
...
Caused by: java.lang.IllegalArgumentException: Could not find build ':'
  	at org.gradle.composite.internal.DefaultIncludedBuildRegistry.getBuild(DefaultIncludedBuildRegistry.java:147)
  	at org.gradle.configurationcache.serialization.codecs.BuildServiceProviderCodec.buildServiceRegistryOf(ProviderCodecs.kt:188)
  	at org.gradle.configurationcache.serialization.codecs.BuildServiceProviderCodec.decode(ProviderCodecs.kt:183)
  	at org.gradle.configurationcache.serialization.codecs.BindingsBackedCodec.decode(BindingsBackedCodec.kt:59)
  	at org.gradle.configurationcache.serialization.codecs.FixedValueReplacingProviderCodec.decodeValue(ProviderCodecs.kt:134)
  	at org.gradle.configurationcache.serialization.codecs.FixedValueReplacingProviderCodec.decodeProvider(ProviderCodecs.kt:116)
  	at org.gradle.configurationcache.serialization.codecs.PropertyCodec.decode(ProviderCodecs.kt:269)
  	at org.gradle.configurationcache.serialization.codecs.BindingsBackedCodec.decode(BindingsBackedCodec.kt:59)
  	at org.gradle.configurationcache.serialization.DefaultReadContext.read(Contexts.kt:259)
  	at org.gradle.configurationcache.serialization.beans.BeanPropertyReaderKt.readPropertyValue(BeanPropertyReader.kt:102)
```

### Context
My goal was to invalidate the configuration cache with the value obtained through the BuildService (with a side effect): the service is starting a web server on a random port, the URL of the server is used to configure some tasks. The build service then stops the server when closed.

The use case is not very compelling, as the lifecycle of the service is unclear in case the cache is actually invalidated - is the service used at fingerprint check time going to be discarded?

### Steps to Reproduce 
See #22338.

--- a/subprojects/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/ConfigurationCacheBuildServiceIntegrationTest.groovy
+++ b/subprojects/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/ConfigurationCacheBuildServiceIntegrationTest.groovy
@@ -21,9 +21,12 @@ import org.gradle.api.Plugin
 import org.gradle.api.Project
 import org.gradle.api.provider.Property
 import org.gradle.api.provider.Provider
+import org.gradle.api.provider.ValueSource
+import org.gradle.api.provider.ValueSourceParameters
 import org.gradle.api.services.BuildService
 import org.gradle.api.services.BuildServiceParameters
 import org.gradle.api.services.ServiceReference
+import org.gradle.api.tasks.Input
 import org.gradle.api.tasks.Internal
 import org.gradle.api.tasks.Optional
 import org.gradle.api.tasks.TaskAction
@@ -38,7 +41,6 @@ import org.gradle.internal.operations.OperationStartEvent
 import org.gradle.tooling.events.FinishEvent
 import org.gradle.tooling.events.OperationCompletionListener
 import org.gradle.tooling.events.task.TaskFinishEvent
-import org.gradle.util.internal.ToBeImplemented
 import spock.lang.IgnoreIf
 import spock.lang.Issue
 
@@ -513,49 +515,206 @@ class ConfigurationCacheBuildServiceIntegrationTest extends AbstractConfiguratio
         outputContains 'probe(classloader2) => 6'
     }
 
-    @ToBeImplemented("https://github.com/gradle/gradle/issues/22337")
-    def "build service can be used as input of value source obtained at configuration time"() {
+    @Issue("https://github.com/gradle/gradle/issues/22337")
+    def "build service can be injected to buildSrc task with #injectionAnnotation"() {
         given:
         def configurationCache = newConfigurationCacheFixture()
-        buildFile("""
-            abstract class EmptyService implements BuildService<${BuildServiceParameters.name}.None> {
-                int getValue() {
-                    return 42
-                }
-            }
 
-            abstract class ServiceValueSource implements ValueSource<Integer, Params> {
-                interface Params extends ValueSourceParameters {
-                    Property<EmptyService> getService()
-                }
-                @Override
-                Integer obtain(){
-                    return parameters.service.get().value
-                }
-            }
-            def serviceProvider = gradle.sharedServices.registerIfAbsent("counter", EmptyService) {}
-            def valueSource = providers.of(ServiceValueSource) {
-                parameters {
-                    service = serviceProvider
-                }
-            }.get()
+        file("buildSrc/build.gradle") << ("""
+            ${constantServiceImpl("ConstantBuildService", "constant")}
+            def serviceProvider = gradle.sharedServices.registerIfAbsent("constant", ConstantBuildService) {}
 
-            task check {
-                doLast {
-                    println "valueSource = " + valueSource
-                }
-            }
+            ${serviceTaskImpl("printValue", injectionAnnotation, propertyType, valueGetter, taskConfiguration)}
+
+            tasks.named("jar") { dependsOn("printValue") }
         """)
 
         when:
-        configurationCacheRun "check"
+        configurationCacheRun()
 
         then:
-        outputContains("valueSource = 42")
-        // TODO(https://github.com/gradle/gradle/issues/22337) A clear error message should be provided at store time
+        // Note that load-after-store doesn't check build fingerprint
         configurationCache.assertStateStored()
-        expect:
-        configurationCacheFails "check"
+
+        when:
+        configurationCacheRun()
+
+        then: "Verify that fingerprint check works"
+        configurationCache.assertStateLoaded()
+
+        where:
+        injectionAnnotation                    | propertyType           | taskConfiguration                                       | valueGetter
+        "${ServiceReference.name}('constant')" | "ConstantBuildService" | ""                                                      | "value.get().value"
+        Internal.name                          | "ConstantBuildService" | "value = serviceProvider; usesService(serviceProvider)" | "value.get().value"
+        Input.name                             | "String"               | "value = serviceProvider.map { it.value }"              | "value.get()"
+    }
+
+    @Issue("https://github.com/gradle/gradle/issues/22337")
+    def "build service can be injected to build task with #injectionAnnotation"() {
+        given:
+        def configurationCache = newConfigurationCacheFixture()
+
+        buildScript("""
+            ${constantServiceImpl("ConstantBuildService", "constant")}
+            def serviceProvider = gradle.sharedServices.registerIfAbsent("constant", ConstantBuildService) {}
+
+            ${serviceTaskImpl("printValue", injectionAnnotation, propertyType, valueGetter, taskConfiguration)}
+        """)
+
+        when:
+        configurationCacheRun("printValue")
+
+        then:
+        // Note that load-after-store doesn't check build fingerprint
+        configurationCache.assertStateStored()
+
+        when:
+        configurationCacheRun("printValue")
+
+        then: "Verify that fingerprint check works"
+        configurationCache.assertStateLoaded()
+
+        where:
+        injectionAnnotation                    | propertyType           | taskConfiguration                                       | valueGetter
+        "${ServiceReference.name}('constant')" | "ConstantBuildService" | ""                                                      | "value.get().value"
+        Internal.name                          | "ConstantBuildService" | "value = serviceProvider; usesService(serviceProvider)" | "value.get().value"
+        Input.name                             | "String"               | "value = serviceProvider.map { it.value }"              | "value.get()"
+    }
+
+    @Issue("https://github.com/gradle/gradle/issues/22337")
+    def "build service cannot be injected to buildSrc task with ValueSource as #injectionAnnotation"() {
+        given:
+        file("buildSrc/build.gradle") << ("""
+            ${constantServiceImpl("ConstantBuildService", "constant")}
+            def serviceProvider = gradle.sharedServices.registerIfAbsent("constant", ConstantBuildService) {}
+
+            ${convertingValueSourceImpl("ConvertingValueSource", valueSourceInputType, "String", conversion)}
+            def valueSource = providers.of(ConvertingValueSource) {
+                parameters.input = $valueSourceInput
+            }
+            ${serviceTaskImpl("printValue", injectionAnnotation, "String", "value.get()", "value = valueSource")}
+
+            tasks.named("jar") { dependsOn("printValue") }
+        """)
+
+        when:
+        configurationCacheFails()
+
+        then:
+        outputContains("Configuration cache entry discarded with 1 problem.")
+        problems.assertFailureHasProblems(failure) {
+            totalProblemsCount = 1
+            // TODO(mlopatkin): Is it possible to figure out a correct location? Report falls back to "Gradle runtime"
+            withProblem("Unknown location: " +
+                "cannot serialize BuildServiceProvider of service 'ConstantBuildService' with name 'constant' used at configuration time as these are not supported with the configuration cache.")
+            problemsWithStackTraceCount = 0
+        }
+
+        where:
+        injectionAnnotation | valueSourceInputType   | valueSourceInput                   | conversion
+        Input.name          | "ConstantBuildService" | "serviceProvider"                  | "parameters.input.get().value"
+        Internal.name       | "ConstantBuildService" | "serviceProvider"                  | "parameters.input.get().value"
+        Input.name          | "String"               | "serviceProvider.map { it.value} " | "parameters.input.get()"
+        Internal.name       | "String"               | "serviceProvider.map { it.value} " | "parameters.input.get()"
+    }
+
+    @Issue("https://github.com/gradle/gradle/issues/22337")
+    def "build service cannot be used in ValueSource if it is obtained at configuration time"() {
+        given:
+        buildScript("""
+            ${constantServiceImpl("ConstantBuildService", "constant")}
+
+            ${convertingValueSourceImpl("ConvertingValueSource", "ConstantBuildService", "String", "parameters.input.get().value")}
+            def valueSource = providers.of(ConvertingValueSource) {
+                parameters.input = gradle.sharedServices.registerIfAbsent("constant", ConstantBuildService) {}
+            }
+
+            println(valueSource.get())
+        """)
+
+        when:
+        configurationCacheFails()
+
+        then:
+        outputContains("Configuration cache entry discarded with 1 problem.")
+        problems.assertFailureHasProblems(failure) {
+            totalProblemsCount = 1
+            withProblem("Build file 'build.gradle': " +
+                "cannot serialize BuildServiceProvider of service 'ConstantBuildService' with name 'constant' used at configuration time as these are not supported with the configuration cache.")
+            problemsWithStackTraceCount = 0
+        }
+    }
+
+    @Issue("https://github.com/gradle/gradle/issues/22337")
+    def "build service can be used in ValueSource if it is used as task input with #injectionAnnotation"() {
+        given:
+        def configurationCache = newConfigurationCacheFixture()
+
+        buildScript("""
+            ${constantServiceImpl("ConstantBuildService", "constant")}
+
+            ${convertingValueSourceImpl("ConvertingValueSource", "ConstantBuildService", "String", "parameters.input.get().value")}
+            def valueSource = providers.of(ConvertingValueSource) {
+                parameters.input = gradle.sharedServices.registerIfAbsent("constant", ConstantBuildService) {}
+            }
+
+            ${serviceTaskImpl("printValue", injectionAnnotation, "String", "value.get()", "value = valueSource")}
+        """)
+
+        when:
+        configurationCacheRun("printValue")
+
+        then:
+        // Note that load-after-store doesn't check build fingerprint
+        configurationCache.assertStateStored()
+
+        when:
+        configurationCacheRun("printValue")
+
+        then: "Verify that fingerprint check works"
+        configurationCache.assertStateLoaded()
+
+        where:
+        injectionAnnotation << [Internal.name, Input.name]
+    }
+
+    static String constantServiceImpl(String serviceClassName, String value) {
+        """
+        abstract class $serviceClassName implements ${BuildService.name}<${BuildServiceParameters.name}.None>{
+            String getValue() {
+                return "$value"
+            }
+        }
+        """
+    }
+
+    static String serviceTaskImpl(String name, String injectionAnnotation, String propertyType, String valueGetter, String taskConfiguration) {
+        """
+        abstract class ServiceTask extends DefaultTask {
+            @$injectionAnnotation
+            abstract Property<$propertyType> getValue()
+
+            @TaskAction def action()  { println("ServiceTask value = \${$valueGetter}") }
+        }
+
+        tasks.register("$name", ServiceTask) {
+            $taskConfiguration
+        }
+        """
+    }
+
+    static String convertingValueSourceImpl(String valueSourceClassName, String inputType, String returnType, String conversion) {
+        """
+        abstract class $valueSourceClassName implements ${ValueSource.name}<$returnType, Params> {
+            interface Params extends ${ValueSourceParameters.name} {
+                Property<$inputType> getInput()
+            }
+
+            @Override $returnType obtain() {
+                return $conversion
+            }
+        }
+        """
     }
 
     @Issue("https://github.com/gradle/gradle/issues/23700")

--- a/subprojects/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/ConfigurationCacheFlowScopeIntegrationTest.groovy
+++ b/subprojects/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/ConfigurationCacheFlowScopeIntegrationTest.groovy
@@ -427,4 +427,37 @@ class ConfigurationCacheFlowScopeIntegrationTest extends AbstractConfigurationCa
     private TestFile scriptFileFor(ScriptTarget target) {
         file(target.fileName)
     }
+
+    def "value source with task result provider cannot be obtained at configuration time"() {
+        given:
+        buildFile("""
+        import org.gradle.api.provider.*
+
+        abstract class ResultSource implements ValueSource<String, Params> {
+            interface Params extends ValueSourceParameters {
+                Property<RequestedTasksResult> getTasksResult();
+            }
+
+            @Override String obtain() {
+                return "I'm not using my parameter"
+            }
+        }
+
+        interface FlowProvidersGetter {
+            @Inject FlowProviders getFlowProviders()
+        }
+
+        def flowProviders = objects.newInstance(FlowProvidersGetter).flowProviders
+
+        providers.of(ResultSource) {
+            parameters.tasksResult = flowProviders.requestedTasksResult
+        }.get()
+
+        """)
+
+        expect:
+        configurationCacheFails()
+        // TODO(mlopatkin) The error message can be improved.
+        failureHasCause(~/Could not isolate value ResultSource(.*) of type ResultSource.Params/)
+    }
 }

--- a/subprojects/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/ConfigurationCacheValueSourceIntegrationTest.groovy
+++ b/subprojects/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/ConfigurationCacheValueSourceIntegrationTest.groovy
@@ -524,4 +524,28 @@ class ConfigurationCacheValueSourceIntegrationTest extends AbstractConfiguration
         "systemProperty" | "-D"
         "gradleProperty" | "-P"
     }
+
+    def "value source with non-serializable output"() {
+        buildFile("""
+        import org.gradle.api.provider.*
+
+        abstract class BrokenSource implements ValueSource<Thread, ValueSourceParameters.None> {
+            @Override Thread obtain() {
+                return new Thread()
+            }
+        }
+
+        providers.of(BrokenSource) {}.get()
+        """)
+
+        when:
+        configurationCacheFails()
+
+        then:
+        problems.assertFailureHasProblems(failure) {
+            totalProblemsCount = 1
+            withProblem("Build file 'build.gradle': cannot serialize object of type 'java.lang.Thread', a subtype of 'java.lang.Thread', as these are not supported with the configuration cache.")
+            problemsWithStackTraceCount = 0
+        }
+    }
 }

--- a/subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/DefaultConfigurationCache.kt
+++ b/subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/DefaultConfigurationCache.kt
@@ -427,7 +427,7 @@ class DefaultConfigurationCache internal constructor(
     fun cacheFingerprintWriterContextFor(outputStream: OutputStream): DefaultWriteContext {
         val (context, codecs) = cacheIO.writerContextFor(outputStream, "fingerprint")
         return context.apply {
-            push(IsolateOwner.OwnerHost(host), codecs.userTypesCodec())
+            push(IsolateOwner.OwnerHost(host), codecs.fingerprintTypesCodec())
         }
     }
 
@@ -490,7 +490,7 @@ class DefaultConfigurationCache internal constructor(
     fun <T> readFingerprintFile(fingerprintFile: ConfigurationCacheStateFile, action: suspend ReadContext.(ConfigurationCacheFingerprintController.Host) -> T): T =
         fingerprintFile.inputStream().use { inputStream ->
             cacheIO.withReadContextFor(inputStream) { codecs ->
-                withIsolate(IsolateOwner.OwnerHost(host), codecs.userTypesCodec()) {
+                withIsolate(IsolateOwner.OwnerHost(host), codecs.fingerprintTypesCodec()) {
                     action(object : ConfigurationCacheFingerprintController.Host {
                         override val valueSourceProviderFactory: ValueSourceProviderFactory
                             get() = host.service()

--- a/subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/fingerprint/ConfigurationCacheFingerprintWriter.kt
+++ b/subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/fingerprint/ConfigurationCacheFingerprintWriter.kt
@@ -426,8 +426,13 @@ class ConfigurationCacheFingerprintWriter(
             }
 
             else -> {
-                sink().write(ValueSource(obtainedValue.uncheckedCast()))
+                // Custom ValueSource implementations may fail to serialize here.
+                // Writing with explicit trace helps to avoid attributing these failures to "Gradle runtime".
+                // TODO(mlopatkin): can we do even better and pinpoint the exact stacktrace in case of failure?
+                val trace = locationFor(null)
+                sink().write(ValueSource(obtainedValue.uncheckedCast()), trace)
                 reportUniqueValueSourceInput(
+                    trace,
                     displayName = when (source) {
                         is Describable -> source.displayName
                         else -> null
@@ -572,16 +577,16 @@ class ConfigurationCacheFingerprintWriter(
     }
 
     private
-    fun reportUniqueValueSourceInput(displayName: String?, typeName: String) {
+    fun reportUniqueValueSourceInput(trace: PropertyTrace, displayName: String?, typeName: String) {
         // We assume different types won't ever produce identical display names
         if (reportedValueSources.add(displayName ?: typeName)) {
-            reportValueSourceInput(displayName, typeName)
+            reportValueSourceInput(trace, displayName, typeName)
         }
     }
 
     private
-    fun reportValueSourceInput(displayName: String?, typeName: String) {
-        reportInput(consumer = null, documentationSection = null) {
+    fun reportValueSourceInput(trace: PropertyTrace, displayName: String?, typeName: String) {
+        reportInput(trace, documentationSection = null) {
             text("value from custom source ")
             reference(typeName)
             displayName?.let {
@@ -718,9 +723,18 @@ class ConfigurationCacheFingerprintWriter(
         documentationSection: DocumentationSection?,
         messageBuilder: StructuredMessage.Builder.() -> Unit
     ) {
+        reportInput(locationFor(consumer), documentationSection, messageBuilder)
+    }
+
+    private
+    fun reportInput(
+        trace: PropertyTrace,
+        documentationSection: DocumentationSection?,
+        messageBuilder: StructuredMessage.Builder.() -> Unit
+    ) {
         host.reportInput(
             PropertyProblem(
-                locationFor(consumer),
+                trace,
                 StructuredMessage.build(messageBuilder),
                 null,
                 documentationSection = documentationSection
@@ -787,7 +801,7 @@ class ConfigurationCacheFingerprintWriter(
             }
         }
 
-        abstract fun write(value: ConfigurationCacheFingerprint)
+        abstract fun write(value: ConfigurationCacheFingerprint, trace: PropertyTrace? = null)
 
         fun inputFile(file: File) =
             InputFile(
@@ -804,8 +818,8 @@ class ConfigurationCacheFingerprintWriter(
         host: Host,
         private val writer: ScopedFingerprintWriter<ConfigurationCacheFingerprint>
     ) : Sink(host) {
-        override fun write(value: ConfigurationCacheFingerprint) {
-            writer.write(value)
+        override fun write(value: ConfigurationCacheFingerprint, trace: PropertyTrace?) {
+            writer.write(value, trace)
         }
 
         fun initScripts(initScripts: List<File>) {
@@ -824,8 +838,8 @@ class ConfigurationCacheFingerprintWriter(
         private val project: Path,
         private val writer: ScopedFingerprintWriter<ProjectSpecificFingerprint>
     ) : Sink(host) {
-        override fun write(value: ConfigurationCacheFingerprint) {
-            writer.write(ProjectSpecificFingerprint.ProjectFingerprint(project, value))
+        override fun write(value: ConfigurationCacheFingerprint, trace: PropertyTrace?) {
+            writer.write(ProjectSpecificFingerprint.ProjectFingerprint(project, value), trace)
         }
     }
 

--- a/subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/fingerprint/ScopedFingerprintWriter.kt
+++ b/subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/fingerprint/ScopedFingerprintWriter.kt
@@ -16,8 +16,10 @@
 
 package org.gradle.configurationcache.fingerprint
 
+import org.gradle.configurationcache.problems.PropertyTrace
 import org.gradle.configurationcache.serialization.DefaultWriteContext
 import org.gradle.configurationcache.serialization.runWriteOperation
+import org.gradle.configurationcache.serialization.withPropertyTrace
 import java.io.Closeable
 
 
@@ -34,9 +36,11 @@ class ScopedFingerprintWriter<T>(
         }
     }
 
-    fun write(value: T) {
+    fun write(value: T, trace: PropertyTrace? = null) {
         synchronized(writeContext) {
-            unsafeWrite(value)
+            withPropertyTrace(trace) {
+                unsafeWrite(value)
+            }
         }
     }
 
@@ -44,6 +48,15 @@ class ScopedFingerprintWriter<T>(
     fun unsafeWrite(value: T?) {
         writeContext.runWriteOperation {
             write(value)
+        }
+    }
+
+    private
+    inline fun withPropertyTrace(trace: PropertyTrace?, block: ScopedFingerprintWriter<T>.() -> Unit) {
+        if (trace != null) {
+            writeContext.withPropertyTrace(trace) { block() }
+        } else {
+            block()
         }
     }
 }

--- a/subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/problems/PropertyProblem.kt
+++ b/subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/problems/PropertyProblem.kt
@@ -36,6 +36,7 @@ enum class DocumentationSection(val anchor: String) {
     NotYetImplementedSourceDependencies("config_cache:not_yet_implemented:source_dependencies"),
     NotYetImplementedJavaSerialization("config_cache:not_yet_implemented:java_serialization"),
     NotYetImplementedTestKitJavaAgent("config_cache:not_yet_implemented:testkit_build_with_java_agent"),
+    NotYetImplementedBuildServiceInFingerprint("config_cache:not_yet_implemented:build_services_in_fingerprint"),
     RequirementsBuildListeners("config_cache:requirements:build_listeners"),
     RequirementsDisallowedTypes("config_cache:requirements:disallowed_types"),
     RequirementsExternalProcess("config_cache:requirements:external_processes"),

--- a/subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/serialization/codecs/Codecs.kt
+++ b/subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/serialization/codecs/Codecs.kt
@@ -119,103 +119,129 @@ class Codecs(
     flowProviders: FlowProviders,
 ) {
     private
-    val userTypesBindings = Bindings.of {
+    val userTypesBindings: Bindings
+    private
+    val fingerprintUserTypesBindings: Bindings
 
-        unsupportedTypes()
+    init {
+        fun makeUserTypeBindings(providersBlock: BindingsBuilder.() -> Unit) = Bindings.of {
+            unsupportedTypes()
 
-        baseTypes()
+            baseTypes()
 
-        bind(HASHCODE_SERIALIZER)
+            bind(HASHCODE_SERIALIZER)
 
-        providerTypes(
-            propertyFactory,
-            filePropertyFactory,
-            valueSourceProviderFactory,
-            buildStateRegistry,
-            flowProviders
-        )
+            providersBlock()
 
-        bind(ListenerBroadcastCodec(listenerManager))
-        bind(LoggerCodec)
+            bind(ListenerBroadcastCodec(listenerManager))
+            bind(LoggerCodec)
 
-        fileCollectionTypes(directoryFileTreeFactory, fileCollectionFactory, artifactSetConverter, fileOperations, fileFactory, patternSetFactory)
+            fileCollectionTypes(directoryFileTreeFactory, fileCollectionFactory, artifactSetConverter, fileOperations, fileFactory, patternSetFactory)
 
-        bind(ApiTextResourceAdapterCodec)
+            bind(ApiTextResourceAdapterCodec)
 
-        bind(ClosureCodec)
-        bind(GroovyMetaClassCodec)
-        bind(SerializedLambdaParametersCheckingCodec)
+            bind(ClosureCodec)
+            bind(GroovyMetaClassCodec)
+            bind(SerializedLambdaParametersCheckingCodec)
 
-        // Dependency management types
-        bind(ArtifactCollectionCodec(calculatedValueContainerFactory, artifactSetConverter))
-        bind(ImmutableAttributesCodec(attributesFactory, managedFactoryRegistry))
-        bind(AttributeContainerCodec(attributesFactory, managedFactoryRegistry))
-        bind(ComponentVariantIdentifierCodec)
-        bind(InitialTransformationNodeCodec(buildOperationExecutor, calculatedValueContainerFactory))
-        bind(ChainedTransformationNodeCodec(buildOperationExecutor, calculatedValueContainerFactory))
-        bind(TransformationStepCodec(inputFingerprinter))
-        bind(TransformationChainCodec())
-        bind(DefaultTransformerCodec(fileLookup, actionScheme))
-        bind(DefaultResolvableArtifactCodec(calculatedValueContainerFactory))
-        bind(TransformStepSpecCodec)
-        bind(PublishArtifactLocalArtifactMetadataCodec)
-        bind(TransformedProjectArtifactSetCodec())
-        bind(TransformedExternalArtifactSetCodec())
-        bind(CalculateArtifactsCodec(calculatedValueContainerFactory))
-        bind(TransformedArtifactCodec(calculatedValueContainerFactory))
-        bind(LocalFileDependencyBackedArtifactSetCodec(instantiator, attributesFactory, calculatedValueContainerFactory))
-        bind(CalculatedValueContainerCodec(calculatedValueContainerFactory))
-        bind(IsolateTransformerParametersNodeCodec(parameterScheme, isolatableFactory, buildOperationExecutor, classLoaderHierarchyHasher, fileCollectionFactory, documentationRegistry))
-        bind(FinalizeTransformDependenciesNodeCodec())
-        bind(ResolveArtifactNodeCodec)
-        bind(WorkNodeActionCodec)
-        bind(CapabilitySerializer())
+            // Dependency management types
+            bind(ArtifactCollectionCodec(calculatedValueContainerFactory, artifactSetConverter))
+            bind(ImmutableAttributesCodec(attributesFactory, managedFactoryRegistry))
+            bind(AttributeContainerCodec(attributesFactory, managedFactoryRegistry))
+            bind(ComponentVariantIdentifierCodec)
+            bind(InitialTransformationNodeCodec(buildOperationExecutor, calculatedValueContainerFactory))
+            bind(ChainedTransformationNodeCodec(buildOperationExecutor, calculatedValueContainerFactory))
+            bind(TransformationStepCodec(inputFingerprinter))
+            bind(TransformationChainCodec())
+            bind(DefaultTransformerCodec(fileLookup, actionScheme))
+            bind(DefaultResolvableArtifactCodec(calculatedValueContainerFactory))
+            bind(TransformStepSpecCodec)
+            bind(PublishArtifactLocalArtifactMetadataCodec)
+            bind(TransformedProjectArtifactSetCodec())
+            bind(TransformedExternalArtifactSetCodec())
+            bind(CalculateArtifactsCodec(calculatedValueContainerFactory))
+            bind(TransformedArtifactCodec(calculatedValueContainerFactory))
+            bind(LocalFileDependencyBackedArtifactSetCodec(instantiator, attributesFactory, calculatedValueContainerFactory))
+            bind(CalculatedValueContainerCodec(calculatedValueContainerFactory))
+            bind(IsolateTransformerParametersNodeCodec(parameterScheme, isolatableFactory, buildOperationExecutor, classLoaderHierarchyHasher, fileCollectionFactory, documentationRegistry))
+            bind(FinalizeTransformDependenciesNodeCodec())
+            bind(ResolveArtifactNodeCodec)
+            bind(WorkNodeActionCodec)
+            bind(CapabilitySerializer())
 
-        bind(DefaultCopySpecCodec(patternSetFactory, fileCollectionFactory, instantiator))
-        bind(DestinationRootCopySpecCodec(fileResolver))
+            bind(DefaultCopySpecCodec(patternSetFactory, fileCollectionFactory, instantiator))
+            bind(DestinationRootCopySpecCodec(fileResolver))
 
-        bind(TaskReferenceCodec)
+            bind(TaskReferenceCodec)
 
-        bind(CachedEnvironmentStateCodec)
+            bind(CachedEnvironmentStateCodec)
 
-        bind(IsolatedManagedValueCodec(managedFactoryRegistry))
-        bind(IsolatedImmutableManagedValueCodec(managedFactoryRegistry))
-        bind(IsolatedJavaSerializedValueSnapshotCodec)
-        bind(IsolatedArrayCodec)
-        bind(IsolatedSetCodec)
-        bind(IsolatedListCodec)
-        bind(IsolatedMapCodec)
-        bind(MapEntrySnapshotCodec)
-        bind(IsolatedEnumValueSnapshotCodec)
-        bind(StringValueSnapshotCodec)
-        bind(IntegerValueSnapshotCodec)
-        bind(FileValueSnapshotCodec)
-        bind(BooleanValueSnapshotCodec)
-        bind(NullValueSnapshotCodec)
+            bind(IsolatedManagedValueCodec(managedFactoryRegistry))
+            bind(IsolatedImmutableManagedValueCodec(managedFactoryRegistry))
+            bind(IsolatedJavaSerializedValueSnapshotCodec)
+            bind(IsolatedArrayCodec)
+            bind(IsolatedSetCodec)
+            bind(IsolatedListCodec)
+            bind(IsolatedMapCodec)
+            bind(MapEntrySnapshotCodec)
+            bind(IsolatedEnumValueSnapshotCodec)
+            bind(StringValueSnapshotCodec)
+            bind(IntegerValueSnapshotCodec)
+            bind(FileValueSnapshotCodec)
+            bind(BooleanValueSnapshotCodec)
+            bind(NullValueSnapshotCodec)
 
-        bind(ServicesCodec())
+            bind(ServicesCodec())
 
-        bind(ProxyCodec)
+            bind(ProxyCodec)
 
-        // Java serialization integration
-        bind(unsupported<Externalizable>(NotYetImplementedJavaSerialization))
-        bind(JavaObjectSerializationCodec(javaSerializationEncodingLookup))
+            // Java serialization integration
+            bind(unsupported<Externalizable>(NotYetImplementedJavaSerialization))
+            bind(JavaObjectSerializationCodec(javaSerializationEncodingLookup))
 
-        bind(BeanSpecCodec)
+            bind(BeanSpecCodec)
+        }
+
+        userTypesBindings = makeUserTypeBindings {
+            providerTypes(
+                propertyFactory,
+                filePropertyFactory,
+                nestedProviderCodec(
+                    valueSourceProviderFactory,
+                    buildStateRegistry,
+                    flowProviders)
+            )
+        }
+
+        fingerprintUserTypesBindings = makeUserTypeBindings {
+            providerTypes(
+                propertyFactory,
+                filePropertyFactory,
+                nestedProviderCodecForFingerprint(
+                    valueSourceProviderFactory,
+                    flowProviders)
+            )
+        }
     }
 
-    fun userTypesCodec(): Codec<Any?> = userTypesBindings.append {
+    private
+    fun Bindings.completeWithReentrantBeanCodec() = append {
         // This protects the BeanCodec against StackOverflowErrors but
         // we can still get them for the other codecs, for instance,
         // with deeply nested Lists, deeply nested Maps, etc.
+        // The reentrant codec is stateful, and cannot be cached because of it.
         bind(reentrant(BeanCodec))
     }.build()
+
+    fun userTypesCodec(): Codec<Any?> = userTypesBindings.completeWithReentrantBeanCodec()
+
+    fun fingerprintTypesCodec(): Codec<Any?> = fingerprintUserTypesBindings.completeWithReentrantBeanCodec()
 
     private
     val internalTypesBindings = Bindings.of {
         baseTypes()
 
-        providerTypes(propertyFactory, filePropertyFactory, valueSourceProviderFactory, buildStateRegistry, flowProviders)
+        providerTypes(propertyFactory, filePropertyFactory, nestedProviderCodec(valueSourceProviderFactory, buildStateRegistry, flowProviders))
         fileCollectionTypes(directoryFileTreeFactory, fileCollectionFactory, artifactSetConverter, fileOperations, fileFactory, patternSetFactory)
 
         bind(TaskInAnotherBuildCodec(includedTaskGraph))
@@ -238,11 +264,8 @@ class Codecs(
     fun BindingsBuilder.providerTypes(
         propertyFactory: PropertyFactory,
         filePropertyFactory: FilePropertyFactory,
-        valueSourceProviderFactory: ValueSourceProviderFactory,
-        buildStateRegistry: BuildStateRegistry,
-        flowProviders: FlowProviders
+        nestedCodec: FixedValueReplacingProviderCodec
     ) {
-        val nestedCodec = FixedValueReplacingProviderCodec(valueSourceProviderFactory, buildStateRegistry, flowProviders)
         bind(ListPropertyCodec(propertyFactory, nestedCodec))
         bind(SetPropertyCodec(propertyFactory, nestedCodec))
         bind(MapPropertyCodec(propertyFactory, nestedCodec))
@@ -251,6 +274,33 @@ class Codecs(
         bind(PropertyCodec(propertyFactory, nestedCodec))
         bind(ProviderCodec(nestedCodec))
     }
+
+    /**
+     * Returns a Codec for Provider implementations suitable for all Provider implementations.
+     */
+    private
+    fun nestedProviderCodec(
+        valueSourceProviderFactory: ValueSourceProviderFactory,
+        buildStateRegistry: BuildStateRegistry,
+        flowProviders: FlowProviders
+    ) = FixedValueReplacingProviderCodec(
+        ValueSourceProviderCodec(valueSourceProviderFactory),
+        BuildServiceProviderCodec(buildStateRegistry),
+        FlowProvidersCodec(flowProviders)
+    )
+
+    /**
+     * Returns a Codec for Provider implementations supported in the fingerprinting context. For example, BuildServiceProviders are not supported.
+     */
+    private
+    fun nestedProviderCodecForFingerprint(
+        valueSourceProviderFactory: ValueSourceProviderFactory,
+        flowProviders: FlowProviders
+    ) = FixedValueReplacingProviderCodec(
+        ValueSourceProviderCodec(valueSourceProviderFactory),
+        UnsupportedFingerprintBuildServiceProviderCodec,
+        FlowProvidersCodec(flowProviders)
+    )
 
     private
     fun BindingsBuilder.fileCollectionTypes(

--- a/subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/serialization/codecs/Codecs.kt
+++ b/subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/serialization/codecs/Codecs.kt
@@ -218,8 +218,7 @@ class Codecs(
                 propertyFactory,
                 filePropertyFactory,
                 nestedProviderCodecForFingerprint(
-                    valueSourceProviderFactory,
-                    flowProviders)
+                    valueSourceProviderFactory)
             )
         }
     }
@@ -294,12 +293,11 @@ class Codecs(
      */
     private
     fun nestedProviderCodecForFingerprint(
-        valueSourceProviderFactory: ValueSourceProviderFactory,
-        flowProviders: FlowProviders
+        valueSourceProviderFactory: ValueSourceProviderFactory
     ) = FixedValueReplacingProviderCodec(
         ValueSourceProviderCodec(valueSourceProviderFactory),
         UnsupportedFingerprintBuildServiceProviderCodec,
-        FlowProvidersCodec(flowProviders)
+        UnsupportedFingerprintFlowProviders
     )
 
     private

--- a/subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/serialization/codecs/ProviderCodecs.kt
+++ b/subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/serialization/codecs/ProviderCodecs.kt
@@ -60,15 +60,15 @@ import org.gradle.internal.build.BuildStateRegistry
  */
 internal
 class FixedValueReplacingProviderCodec(
-    valueSourceProviderFactory: ValueSourceProviderFactory,
-    buildStateRegistry: BuildStateRegistry,
-    flowProviders: FlowProviders
+    valueSourceProviderCodec: Codec<ValueSourceProvider<*, *>>,
+    buildServiceProviderCodec: Codec<BuildServiceProvider<*, *>>,
+    flowProvidersCodec: Codec<RequestedTasksResultProvider>,
 ) {
     private
     val providerWithChangingValueCodec = Bindings.of {
-        bind(ValueSourceProviderCodec(valueSourceProviderFactory))
-        bind(BuildServiceProviderCodec(buildStateRegistry))
-        bind(FlowProvidersCodec(flowProviders))
+        bind(valueSourceProviderCodec)
+        bind(buildServiceProviderCodec)
+        bind(flowProvidersCodec)
         bind(BeanCodec)
     }.build()
 

--- a/subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/serialization/codecs/UnsupportedTypesCodecs.kt
+++ b/subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/serialization/codecs/UnsupportedTypesCodecs.kt
@@ -55,8 +55,10 @@ import org.gradle.api.tasks.SourceSet
 import org.gradle.api.tasks.SourceSetContainer
 import org.gradle.api.tasks.TaskContainer
 import org.gradle.api.tasks.TaskDependency
+import org.gradle.configurationcache.flow.RequestedTasksResultProvider
 import org.gradle.configurationcache.problems.DocumentationSection
 import org.gradle.configurationcache.serialization.Codec
+import org.gradle.configurationcache.serialization.IsolateContext
 import org.gradle.configurationcache.serialization.ReadContext
 import org.gradle.configurationcache.serialization.WriteContext
 import org.gradle.configurationcache.serialization.logUnsupported
@@ -159,5 +161,27 @@ object UnsupportedFingerprintBuildServiceProviderCodec : Codec<BuildServiceProvi
     override suspend fun ReadContext.decode(): BuildServiceProvider<*, *>? {
         logUnsupported("deserialize", BuildServiceProvider::class, documentationSection = DocumentationSection.NotYetImplementedBuildServiceInFingerprint)
         return null
+    }
+}
+
+
+internal
+object UnsupportedFingerprintFlowProviders : Codec<RequestedTasksResultProvider> {
+    override suspend fun WriteContext.encode(value: RequestedTasksResultProvider) {
+        logUnsupported("serialize")
+    }
+
+    override suspend fun ReadContext.decode(): RequestedTasksResultProvider? {
+        logUnsupported("deserialize")
+        return null
+    }
+
+    private
+    fun IsolateContext.logUnsupported(action: String) {
+        logUnsupported(action) {
+            text(" ")
+            reference(RequestedTasksResultProvider::class)
+            text(" used at configuration time")
+        }
     }
 }

--- a/subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/serialization/codecs/UnsupportedTypesCodecs.kt
+++ b/subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/serialization/codecs/UnsupportedTypesCodecs.kt
@@ -50,10 +50,16 @@ import org.gradle.api.initialization.Settings
 import org.gradle.api.invocation.Gradle
 import org.gradle.api.publish.Publication
 import org.gradle.api.services.BuildService
+import org.gradle.api.services.internal.BuildServiceProvider
 import org.gradle.api.tasks.SourceSet
 import org.gradle.api.tasks.SourceSetContainer
 import org.gradle.api.tasks.TaskContainer
 import org.gradle.api.tasks.TaskDependency
+import org.gradle.configurationcache.problems.DocumentationSection
+import org.gradle.configurationcache.serialization.Codec
+import org.gradle.configurationcache.serialization.ReadContext
+import org.gradle.configurationcache.serialization.WriteContext
+import org.gradle.configurationcache.serialization.logUnsupported
 import org.gradle.configurationcache.serialization.unsupported
 import org.gradle.internal.scripts.GradleScript
 import org.gradle.internal.service.DefaultServiceRegistry
@@ -134,4 +140,24 @@ fun BindingsBuilder.unsupportedTypes() {
 
     // Gradle implementation types
     bind(unsupported<DefaultServiceRegistry>())
+}
+
+
+internal
+object UnsupportedFingerprintBuildServiceProviderCodec : Codec<BuildServiceProvider<*, *>> {
+    override suspend fun WriteContext.encode(value: BuildServiceProvider<*, *>) {
+        logUnsupported("serialize",
+            documentationSection = DocumentationSection.NotYetImplementedBuildServiceInFingerprint) {
+            text(" BuildServiceProvider of service ")
+            reference(value.serviceDetails.implementationType)
+            text(" with name ")
+            reference(value.serviceDetails.name)
+            text(" used at configuration time")
+        }
+    }
+
+    override suspend fun ReadContext.decode(): BuildServiceProvider<*, *>? {
+        logUnsupported("deserialize", BuildServiceProvider::class, documentationSection = DocumentationSection.NotYetImplementedBuildServiceInFingerprint)
+        return null
+    }
 }

--- a/subprojects/docs/src/docs/userguide/running-builds/configuration_cache.adoc
+++ b/subprojects/docs/src/docs/userguide/running-builds/configuration_cache.adoc
@@ -1094,3 +1094,13 @@ include::sample[dir="snippets/configurationCache/topLevel/kotlin",files="build.g
 <2> Define the variable in a smaller scope.
 
 See link:{gradle-issues}22879[gradle/gradle#22879].
+
+[[config_cache:not_yet_implemented:build_services_in_fingerprint]]
+=== Using build services to invalidate the configuration cache
+
+Currently, it is not possible to use a `BuildServiceProvider` or provider derived of it with `map` or `flatMap` as a parameter for the `ValueSource`, if the value of the `ValueSource` is accessed at configuration time.
+The same applies such `ValueSource` is obtained in a task that executes as part of configuration phase, for example tasks of the `buildSrc` build or included builds contributing plugins.
+Note that using a `@ServiceReference` or storing `BuildServiceProvider` in an `@Internal`-annotated property of a task is safe.
+Generally speaking, this limitation makes it impossible to use a `BuildService` to invalidate the configuration cache.
+
+See link:{gradle-issues}24085[gradle/gradle#24085].

